File: D04/Golang/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)

const XmasSpan = 3

var XmasExtents = []Point{
	{0, XmasSpan},
	{0, -XmasSpan},
	{-XmasSpan, 0},
	{XmasSpan, 0},
	{XmasSpan, XmasSpan},
	{XmasSpan, -XmasSpan},
	{-XmasSpan, -XmasSpan},
	{-XmasSpan, XmasSpan},
}

var XmasCrossExtents = []Point{
	{1, 1},
	{1, -1},
	{-1, -1},
	{-1, 1},
}

type Table struct {
	cells   [][]rune
	numRows int
	numCols int
}

type Point struct {
	x int
	y int
}

type Bound struct {
	startX int
	endX   int
	startY int
	endY   int
}

func (c Bound) String() string {
	return fmt.Sprintf("(%d, %d) -> (%d, %d)", c.startX, c.startY, c.endX, c.endY)
}

func (t Table) String() string {
	var table string
	for _, row := range t.cells {
		var rowStr string
		for _, ch := range row {
			rowStr += string(ch)
		}
		table += rowStr + "\n"
	}
	return table
}

func processTable(input string) Table {
	input = strings.Trim(input, "\n")
	var table Table
	var rows [][]rune
	for _, line := range strings.Split(input, "\n") {
		var row []rune
		for _, ch := range line {
			row = append(row, ch)
		}
		table.numCols = len(row)
		rows = append(rows, row)
	}
	table.numRows = len(rows)
	table.cells = rows
	return table
}

func scanTable(table Table, x, y int) []Bound {
	var xmasCount []Bound

	if table.cells[x][y] != 'X' {
		return nil
	}

	for _, extent := range XmasExtents {
		maxX := x + extent.x
		maxY := y + extent.y
		if !(maxX >= 0 && maxX < table.numRows && maxY >= 0 && maxY < table.numCols) {
			continue
		}

		var xmasCheck string
		for step := 0; step < 4; step++ {
			curX := x
			if maxX-x == XmasSpan {
				curX += step
			} else if maxX-x == -XmasSpan {
				curX -= step
			}

			curY := y
			if maxY-y == XmasSpan {
				curY += step
			} else if maxY-y == -XmasSpan {
				curY -= step
			}
			xmasCheck += string(table.cells[curX][curY])
		}
		if xmasCheck == "XMAS" {
			xmasCount = append(xmasCount, Bound{
				startX: x,
				startY: y,
				endX:   maxX,
				endY:   maxY,
			})
		}
	}
	return xmasCount
}

func scanTableCross(table Table, x, y int) *Point {
	if table.cells[x][y] != 'A' {
		return nil
	}

	for _, extent := range XmasCrossExtents {
		maxX := x + extent.x
		maxY := y + extent.y
		if !(maxX >= 0 && maxX < table.numRows && maxY >= 0 && maxY < table.numCols) {
			return nil
		}
	}

	tl := table.cells[x-1][y-1]
	tr := table.cells[x-1][y+1]
	bl := table.cells[x+1][y-1]
	br := table.cells[x+1][y+1]

	m1 := string(tl) + "A" + string(br)
	m2 := string(bl) + "A" + string(tr)

	if (m1 == "MAS" || m1 == "SAM") && (m2 == "MAS" || m2 == "SAM") {
		return &Point{x: x, y: y}
	}

	return nil
}

func SolvePartOne(input string) int {
	table := processTable(input)

	var total int

	for idx, row := range table.cells {
		for jdx := range row {
			total += len(scanTable(table, idx, jdx))
		}
	}

	return total
}

func SolvePartTwo(input string) int {
	table := processTable(input)

	var total int

	for idx, row := range table.cells {
		for jdx := range row {
			extent := scanTableCross(table, idx, jdx)
			if extent != nil {
				total += 1
			}
		}
	}

	return total
}

func main() {
	data, _ := os.ReadFile("../input.txt")

	a1 := SolvePartOne(string(data))

	fmt.Println(a1)

	a2 := SolvePartTwo(string(data))

	fmt.Println(a2)
}

File: D04/Rust/src/main.rs
static XMAS: &str = "XMAS";

const DIRECTIONS: [(isize, isize); 8] = [
    (0, 1),
    (0, -1),
    (-1, 0),
    (1, 0),
    (1, 1),
    (1, -1),
    (-1, -1),
    (-1, 1),
];

#[derive(Default)]
struct Table {
    cells: Vec<Vec<char>>,
    row_count: usize,
    col_count: usize,
}

impl Table {
    pub fn new(input: &str) -> Self {
        let cells: Vec<Vec<char>> = input.lines().map(|line| line.chars().collect()).collect();

        Table {
            row_count: cells.len(),
            col_count: cells.first().map_or(0, |row| row.len()),
            cells,
        }
    }

    fn coordinate_map(&self) -> impl Iterator<Item = (usize, usize)> + '_ {
        (0..self.row_count).flat_map(|x| (0..self.col_count).map(move |y| (x, y)))
    }
}

fn scan_table(table: &Table, x: usize, y: usize) -> usize {
    if table.cells[x][y] != 'X' {
        0
    } else {
        DIRECTIONS
            .iter()
            .filter(|(dx, dy)| {
                XMAS.chars().enumerate().all(|(step, ch)| {
                    let nx = x as isize + dx * step as isize;
                    let ny = y as isize + dy * step as isize;

                    nx >= 0
                        && nx < table.row_count as isize
                        && ny >= 0
                        && ny < table.col_count as isize
                        && table.cells[nx as usize][ny as usize] == ch
                })
            })
            .count()
    }
}

fn scan_table_cross(table: &Table, x: usize, y: usize) -> bool {
    if table.cells[x][y] != 'A'
        || x == 0
        || y == 0
        || x >= table.row_count - 1
        || y >= table.col_count - 1
    {
        return false;
    }

    let corners = [
        table.cells[x - 1][y - 1], // tl
        table.cells[x - 1][y + 1], // tr
        table.cells[x + 1][y - 1], // bl
        table.cells[x + 1][y + 1], // br
    ];

    let diag1 = format!("{}A{}", corners[0], corners[3]);
    let diag2 = format!("{}A{}", corners[1], corners[2]);

    (diag1 == "MAS" || diag1 == "SAM") && (diag2 == "MAS" || diag2 == "SAM")
}

fn solve_part_one(input: &str) -> usize {
    let table = Table::new(input);

    table
        .coordinate_map()
        .fold(0, |acc, (x, y)| acc + scan_table(&table, x, y))
}

fn solve_part_two(input: &str) -> usize {
    let table = Table::new(input);

    table
        .coordinate_map()
        .filter(|&(x, y)| scan_table_cross(&table, x, y))
        .count()
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let data = std::fs::read_to_string("../input.txt")?;

    let p1 = solve_part_one(&data);

    println!("{p1}");

    let p2 = solve_part_two(&data);

    println!("{p2}");

    Ok(())
}

#[cfg(test)]
mod tests {
    static SMALL: &str = r"..X...
.SAMX.
.A..A.
XMAS.S
.X....";

    static SAMPLE: &str = r"MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX";

    #[test]
    fn test_solve_small_one() {
        let expected = 4;

        let actual = super::solve_part_one(SMALL);

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_solve_small_two() {
        let expected = 0;

        let actual = super::solve_part_two(SMALL);

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_solve_one() {
        let expected = 18;

        let actual = super::solve_part_one(SAMPLE);

        assert_eq!(expected, actual);
    }

    #[test]
    fn test_solve_two() {
        let expected = 9;

        let actual = super::solve_part_two(SAMPLE);

        assert_eq!(expected, actual);
    }
}
